examples/vision/convmixer.py

```diff
@@ -0,0 +1,319 @@
+small patches from the input images, linearly project them, and then apply the
+Transformer ([Vaswani et al.](https://arxiv.org/abs/1706.03762)) blocks. The application
+of ViTs to image recognition tasks is quickly becoming a promising area of research
+because ViTs eliminate the need to have strong inductive biases such convolutions for
+modeling locality. This presents them as a general computation primititive capable of
+learning just from the training data with as minimal inductive priors as possible. ViTs
+yield tremendous downstream performance when trained with proper regularization, data
+augmentation, and relatively larger datasets. 
+
+In the [Patches Are All You Need](https://openreview.net/pdf?id=TVHS5Y4dNvM) paper, (at
+the time of writing it is a submission to the ICLR 2022 conference), the authors extend
+the idea of using patches to train an all-convolutional network and demonstrate
+competitive results. Their architecture namely **ConvMixer** uses recipes from the recent
+isotrophic architectures like ViT, MLP-Mixer
+([Tolstikhin et al.](https://arxiv.org/abs/2105.01601)) and these include using same
+depth and resolution across different layers in the network, residual connections,
+and so on.
+
+In this example, we will implement the ConvMixer model and demonstrate its performance on
+the CIFAR-10 dataset.
+
+To use the AdamW optimizer, we need to install TensorFlow Addons.
+
+```shell
+pip install -U -q tensorflow-addons
+```
+"""
+
+"""
+## Imports
+"""
+
+from tensorflow.keras import layers
+from tensorflow import keras
+
+import matplotlib.pyplot as plt
+import tensorflow_addons as tfa
+import tensorflow as tf
+
+"""
+## Hyperparameters
+
+For the purpose of this example, we will train the model for only 10 epochs. To focus on
+the core ideas of ConvMixer, we will not use other training-specific elements like
+RandAugment ([Cubuk et al.](https://arxiv.org/abs/1909.13719)). If you are interested to
+learn those details, please refer to the
+[original paper](https://openreview.net/pdf?id=TVHS5Y4dNvM). 
+"""
+
+learning_rate = 0.001
+weight_decay = 0.0001
+batch_size = 128
+num_epochs = 10
+
+"""
+## Load the CIFAR-10 dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+val_split = 0.1
+
+val_indices = int(len(x_train) * val_split)
+new_x_train, new_y_train = x_train[val_indices:], y_train[val_indices:]
+x_val, y_val = x_train[:val_indices], y_train[:val_indices]
+
+print(f"Training data samples: {len(new_x_train)}")
+print(f"Validation data samples: {len(x_val)}")
+print(f"Test data samples: {len(x_test)}")
+
+"""
+## Prepare `tf.data.Dataset` objects
+
+Our data augmentation pipeline is different from what the authors used for the CIFAR-10
+dataset which is fine for the purpose of the example. 
+"""
+
+image_size = 32
+auto = tf.data.AUTOTUNE
+
+data_augmentation = keras.Sequential(
+    [
+        layers.Rescaling(scale=1.0 / 255),
```

Comment:
Consider moving this layer to be the first layer in the model, so that the model is able to process raw images (and therefore end users don't need to be aware of the particular preprocessing scheme)